crates/ra_vfs/src/lib.rs

```diff
@@ -16,52 +16,61 @@
 mod io;
 
 use std::{
-    fmt,
-    mem,
-    thread,
     cmp::Reverse,
+    fmt, fs, mem,
     path::{Path, PathBuf},
-    ffi::OsStr,
     sync::Arc,
-    fs,
+    thread,
 };
 
-use rustc_hash::{FxHashMap, FxHashSet};
-use relative_path::RelativePathBuf;
 use crossbeam_channel::Receiver;
-use walkdir::DirEntry;
-use thread_worker::WorkerHandle;
-use ra_arena::{Arena, RawId, impl_arena_id};
+use ra_arena::{impl_arena_id, Arena, RawId};
+use relative_path::{Component, RelativePath, RelativePathBuf};
+use rustc_hash::{FxHashMap, FxHashSet};
 
 pub use crate::io::TaskResult as VfsTask;
+use io::{Task, TaskResult, WatcherChange, WatcherChangeData, Worker};
 
 /// `RootFilter` is a predicate that checks if a file can belong to a root. If
 /// several filters match a file (nested dirs), the most nested one wins.
-struct RootFilter {
+pub(crate) struct RootFilter {
     root: PathBuf,
-    file_filter: fn(&Path) -> bool,
+    filter: fn(&Path, &RelativePath) -> bool,
 }
 
 impl RootFilter {
     fn new(root: PathBuf) -> RootFilter {
         RootFilter {
             root,
-            file_filter: has_rs_extension,
+            filter: default_filter,
         }
     }
     /// Check if this root can contain `path`. NB: even if this returns
     /// true, the `path` might actually be conained in some nested root.
-    fn can_contain(&self, path: &Path) -> Option<RelativePathBuf> {
-        if !(self.file_filter)(path) {
+    pub(crate) fn can_contain(&self, path: &Path) -> Option<RelativePathBuf> {
+        let rel_path = path.strip_prefix(&self.root).ok()?;
+        let rel_path = RelativePathBuf::from_path(rel_path).ok()?;
+        if !(self.filter)(path, rel_path.as_relative_path()) {
             return None;
         }
-        let path = path.strip_prefix(&self.root).ok()?;
-        RelativePathBuf::from_path(path).ok()
+        Some(rel_path)
     }
 }
 
-fn has_rs_extension(p: &Path) -> bool {
-    p.extension() == Some(OsStr::new("rs"))
+pub(crate) fn default_filter(path: &Path, rel_path: &RelativePath) -> bool {
+    if path.is_dir() {
+        for (i, c) in rel_path.components().enumerate() {
+            if let Component::Normal(c) = c {
+                // hardcoded for now
```

Comment:
Perhaps add a `TODO: ` in front of this comment, to make it easier to find things that require some rethinking later?

Reply:
good idea